Mathlib/RingTheory/PowerSeries/Eval.lean

```diff
@@ -0,0 +1,345 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.PowerSeries.Trunc
+import Mathlib.RingTheory.PowerSeries.Inverse
+import Mathlib.RingTheory.AdicCompletion.Noetherian
+import Mathlib.Algebra.Polynomial.Div
+
+/-!
+
+# Evaluation of power series
+
+## Main definition
+- `PowerSeries.eval`:
+  `p.eval x` is the evaluation of `p : R⟦X⟧` at `x`,
+  and `0` if the former is not well defined.
+- `PowerSeries.evalAlgHom`: The evaluation map as an algebraic homomorphism.
+
+-/
+
+namespace PowerSeries
+
+variable {R S : Type*} [CommRing R] [CommRing S]
+variable (x : R)
+
+abbrev _root_.IsHausdorffAt : Prop := IsHausdorff (.span {x}) R
+abbrev _root_.IsPrecompleteAt : Prop := IsPrecomplete (.span {x}) R
+abbrev _root_.IsAdicCompleteAt : Prop := IsAdicComplete (.span {x}) R
+
+variable {x} in
+lemma _root_.isHausdorffAt_iff :
+    IsHausdorffAt x ↔ ⨅ i, Ideal.span {x ^ i} = ⊥ := by
+  simp [isHausdorff_iff, SModEq.zero, ← le_bot_iff, SetLike.le_def, Ideal.span_singleton_pow]
+
+/--
+`p : R⟦X⟧` evaluates to `a` at `x` (`p.EvalEq x a`) if `∑ᵢⁿ⁻¹ pᵢxⁱ ≡ a (mod xⁿ)` for all `n`.
+-/
+def EvalEq (p : R⟦X⟧) (a : R) : Prop := ∀ i, x ^ i ∣ (p.trunc i).eval x - a
```

Comment:
Does `EvalEq p` necessarily uniquely identify `a`? If not, then perhaps this should be called `HasEval p` instead, to match `HasSum`, `HasDerivAt` etc.

Reply:
Only under some separability conditions. I'll change the name.

Reply:
This is probably the only place this PR is more general than #15019; that PR uses a [T2Space assumption](https://github.com/leanprover-community/mathlib4/pull/15019/files#diff-b53af0fbdc571aa34f4b4305e167fd8ee826328acc17b1c26a80ec9fff585fb4R260-R272). (In `eval₂` where T2Space is not assumed, it makes an arbitrary choice.)

Reply:
Oh I'm not aware of that PR. Their results are definitely more general. We probably don't care about evaluation at non-hausdorff rings anyway as the value is not unique.

A slight drawback is that they require a topology on `R` but in some cases we might not want to choose a topology (e.g. on $$\mathbb{Z}_p[[X]]$$ there's a $$p$$-adic one and a $$X$$-adic one). Maybe we should keep both?

Reply:
If you endow `R` with the `x`-adic topology, then you can use our PR to evaluate power series at `x`. The T2Space assumption is more or less necessary to have a well defined sum (given by `choose`, from `hasSum`), but the continuity property should not depend on that, as observed by @ADedecker .